[SIL] Add `GraphOperationInst`.

dan-zheng opened the pull request:
Add `GraphOperationInst` to represent graph operations.

A graph operation may have:
- Zero or more SILValue operands
- Zero or more attributes (consisting of a name and a `SymbolicValue` constant value)
- One or more result types

`GraphOperationInst` will replace `BuiltinInst` as the representation for
TensorFlow operations in the TFDeabstraction/TFPartition passes.

Todos:
 - @lattner: Edit deabstraction/partitioning to use `GraphOperationInst`.
 - Low priority: Implement SIL serialization/deserialization for `GraphOperationInst`.

dan-zheng committed:
[SIL] Add `GraphOperationInst`.

Add `GraphOperationInst` to represent graph operations.

A graph operation may have:
- Multiple SILValue operands
- Multiple attributes (consisting of a name and a `SymbolicValue` constant value)
- Multiple result types

`GraphOperationInst` will replace `BuiltinInst` as the representation for
TensorFlow operations in the TFDeabstraction/TFPartition passes.

Todos:
 - @lattner: Edit deabstraction/partitioning to use `GraphOperationInst`.
 - Implement SIL serialization/deserialization for `GraphOperationInst`.

dan-zheng committed:
Remove extraneous comment.

dan-zheng committed:
Add verification for `GraphOperationInst`.

rxwei commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -7978,6 +7982,130 @@ inline DestructureTupleInst *DestructureTupleResult::getParent() {
+      for (unsigned i : indices(arguments)) {
+        new (&allOperands[i]) Operand(this, arguments[i]);
+      }
+      memcpy(getAttributes().data(), attrs.begin(),
```
`std::copy`?

dan-zheng replied:
Done.

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -885,6 +887,74 @@ void SILParser::convertRequirements(SILFunction *F,
+    // StringRef rawString = P.Tok.getText().drop_front().drop_back();
+    // value = SymbolicValue::getStringValue(rawString, allocator);
+    // return false;
+    llvm_unreachable("`SymbolicValue.getStringValue` is unimplemented");
```
Rather than making it trap I would just emit a diagnostic.

dan-zheng replied:
Done.

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -2767,6 +2837,84 @@ bool SILParser::parseSILInstruction(SILBuilder &B) {
+      if (parseSILType(temp))
+        return true;
+      resultTypes.push_back(temp);
+      if (!P.Tok.is(tok::comma))
```
Replace these three lines with:

```cpp
if (!P.consumeIf(tok::comma))
  break;
```

dan-zheng replied:
Done, thanks.

rxwei commented on lib/SIL/SILInstructions.cpp:
```diff
@@ -2463,3 +2463,25 @@ DestructureTupleInst *DestructureTupleInst::create(SILModule &M,
+    ArrayRef<SILType> resultTypes) {
+  llvm::SmallVector<ValueOwnershipKind, 4> resultOwnerships;
+  for (auto resultType : resultTypes) {
+    if (resultType.isTrivial(M)) {
```
Use a ternary operator.

dan-zheng replied:
Done.

rxwei commented on lib/SIL/SILInstructions.cpp:
```diff
@@ -2463,3 +2463,25 @@ DestructureTupleInst *DestructureTupleInst::create(SILModule &M,
+  }
+
+  unsigned size =
+    totalSizeToAlloc<MultipleValueInstruction *, GraphOperationResult, Operand, GraphOperationAttribute>(
```
80 cols

dan-zheng replied:
Done.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -1176,6 +1176,78 @@ class SILPrinter : public SILInstructionVisitor<SILPrinter> {
+  // SWIFT_ENABLE_TENSORFLOW
+  void visitSymbolicValue(SymbolicValue v) {
+    switch (v.getKind()) {
+      case swift::SymbolicValue::Integer:
```
`case` should be aligned with `switch`.

dan-zheng replied:
Done.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -1176,6 +1176,78 @@ class SILPrinter : public SILInstructionVisitor<SILPrinter> {
+        *this << v.getIntegerValue();
+        return;
+      case swift::SymbolicValue::Float: {
+        SmallString<12> decimal;
```
Why is it 12? Is it an arbitrary number?

dan-zheng replied:
`SmallString<12>` is what's used in [`visitFloatLiteralInst`](https://github.com/apple/swift/blob/9a6adcec9d09481091de9a10d8d45b4445a26995/lib/SIL/SILPrinter.cpp#L1210), I just copied that.
If you'd prefer a more auspicious number, let me know.

rxwei commented on lib/SIL/SILVerifier.cpp:
```diff
@@ -1185,7 +1185,13 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
+  void checkGraphOperationInst(GraphOperationInst *GI) {
+    for (auto attr : GI->getAttributes()) {
+      require(attr.value.isConstant(), "Invalid graph operation attribute");
+    }
```
No need for `{}`

dan-zheng replied:
Done.

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -2767,6 +2837,84 @@ bool SILParser::parseSILInstruction(SILBuilder &B) {
+    P.consumeToken(tok::string_literal);
+
+    // Parse graph operation arguments.
+    if (P.Tok.getKind() != tok::l_paren) {
```
`if (P.Tok.isNot(tok::l_paren))`

dan-zheng replied:
Done.

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -2767,6 +2837,84 @@ bool SILParser::parseSILInstruction(SILBuilder &B) {
+  // SWIFT_ENABLE_TENSORFLOW
+  case SILInstructionKind::GraphOperationInst: {
+    // Parse graph operation name.
+    if (P.Tok.getKind() != tok::string_literal) {
```
`if (P.Tok.isNot(tok::string_literal))`

dan-zheng replied:
Done.

rxwei commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -7978,6 +7982,130 @@ inline DestructureTupleInst *DestructureTupleResult::getParent() {
+                                            resultOwnerships),
+    Name(name), NumOperands(arguments.size()), NumAttributes(attrs.size()) {
+      auto allOperands = getAllOperands();
+      for (unsigned i : indices(arguments)) {
```
Remove `{}`?

dan-zheng replied:
Done.

dan-zheng committed:
Minor cleanup for `GraphOperationInst`.

Address comments from @rxwei.

dan-zheng committed:
Add SWIFT_ENABLE_TENSORFLOW comments.

dan-zheng committed:
Minor fix: return true in `parseSymbolicValue` on error.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Change `getAttribute` to take a `StringRef` instead of an `Identifier`.

`StringRef` is easier to use for callers.

lattner commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -7978,6 +7982,131 @@ inline DestructureTupleInst *DestructureTupleResult::getParent() {
+
+/// SWIFT_ENABLE_TENSORFLOW
+/// An graph operation attribute.
+/// Currently, attributes have a name and a constant value.
```
Not a big deal, but you can drop "Currently,".

dan-zheng replied:
Done.

lattner commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -7978,6 +7982,131 @@ inline DestructureTupleInst *DestructureTupleResult::getParent() {
+      auto allOperands = getAllOperands();
+      for (unsigned i : indices(arguments))
+        new (&allOperands[i]) Operand(this, arguments[i]);
+      std::copy(attrs.begin(), attrs.end(), getAttributes().data());
```
I think this should use uninitialized_copy ?

lattner replied:
Because the destination is uninitialized memory

dan-zheng replied:
Done.

lattner commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -885,6 +887,75 @@ void SILParser::convertRequirements(SILFunction *F,
+    APInt intValue;
+    if (SP.parseInteger(intValue, diag::sil_expected_graph_op_attr_value))
+      return true;
+    value = SymbolicValue::getInteger(intValue, allocator);
```
I know I told you that we don't need types, but I think this is going to be problematic.  We actually do need to know the data type for integers and fp values :-(  Sorry about that.

How about syntax like:
[i64 42, f64 3.0]

Where you just handle i[digit]+  and f[digit]+, and special case f32/f64 for now?

dan-zheng replied:
How about this syntax?
`[name1 42 : i64, name2 3.0 : f64]`

dan-zheng replied:
Is it necessary to keep track of signed-ness (uint vs int)?

lattner replied:
Using "value : type" makes sense, but FYI the constants here have a different type system than Swift does, so please don't try to reuse the existing productions.

lattner commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -885,6 +887,75 @@ void SILParser::convertRequirements(SILFunction *F,
+  // Handle aggregate literals.
+  if (P.consumeIf(tok::l_square)) {
+    SmallVector<SymbolicValue, 8> elements;
+    while (true) {
```
This doesn't look quite right.  It was accept  [123,].  Comma separated lists are annoying, and it is doubly annoying that the SIL parser doesn't have a helper for this.  Please search for tok::comma to see how other loops in the SIL parser are implemented and follow one of their patterns, thanks!

lattner replied:
Actually, it looks like Parser::parseList might be able to do this.  SILParser has a "Parser" member named P.

dan-zheng replied:
Done. I used `parseList` here and in the other places.

lattner commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -2767,6 +2838,83 @@ bool SILParser::parseSILInstruction(SILBuilder &B) {
+    }
+    StringRef rawString = P.Tok.getText().drop_front().drop_back();
+    Identifier name = P.Context.getIdentifier(rawString);
+    P.consumeToken(tok::string_literal);
```
Can you use P.parseIdentifier for this?

dan-zheng replied:
I don't believe so, because `parseIdentifier` doesn't handle quotations that surround a `string_literal` token.

I copied this from the [code for parsing `StringLiteralInst`](https://github.com/apple/swift/blob/9a6adcec9d09481091de9a10d8d45b4445a26995/lib/ParseSIL/ParseSIL.cpp#L2467).
It also uses `P.Tok.getText().drop_front().drop_back()`.

lattner replied:
Ok!

lattner commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -2767,6 +2838,83 @@ bool SILParser::parseSILInstruction(SILBuilder &B) {
+    SmallVector<SILValue, 4> arguments;
+    P.consumeToken(tok::l_paren);
+    while (true) {
+      if (P.consumeIf(tok::r_paren))
```
Same thing about the comma separated list thing.

dan-zheng replied:
Done.

lattner commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -2767,6 +2838,83 @@ bool SILParser::parseSILInstruction(SILBuilder &B) {
+    // Parse optional graph operation attributes.
+    SmallVector<GraphOperationAttribute, 4> attributes;
+    SourceLoc lastLoc = P.getEndOfPreviousLoc();
+    if (P.consumeIf(tok::l_square)) {
```
again :)

dan-zheng replied:
Done.

lattner commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -1176,6 +1176,78 @@ class SILPrinter : public SILInstructionVisitor<SILPrinter> {
+  // SWIFT_ENABLE_TENSORFLOW
+  void visitSymbolicValue(SymbolicValue v) {
+    switch (v.getKind()) {
+    case swift::SymbolicValue::Integer:
```
I don't think the swift:: namespaces are required.

dan-zheng replied:
Done, sorry about that.
`swift::` was produced by Xcode when it auto-generated enum cases.

lattner commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -1176,6 +1176,78 @@ class SILPrinter : public SILInstructionVisitor<SILPrinter> {
+      return;
+    }
+    case swift::SymbolicValue::String:
+      // TODO: Uncomment when `getStringValue` is implemented.
```
Want to take this on as your next project? :-)

dan-zheng replied:
Sure!

lattner commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -1176,6 +1176,78 @@ class SILPrinter : public SILInstructionVisitor<SILPrinter> {
+  }
+
+  // SWIFT_ENABLE_TENSORFLOW
+  void visitGraphOperationInst(GraphOperationInst *GI) {
```
very nice!

lattner commented on test/TensorFlow/graph_op_inst.sil:
```diff
@@ -0,0 +1,61 @@
+
+sil @chained_op_test : $@convention(thin) (Tensor<Float>, Tensor<Float>) -> Tensor<Float> {
+bb0(%0 : $Tensor<Float>, %1 : $Tensor<Float>):
+  %2 = graph_op "tf.Add"(%0 : $Tensor<Float>, %1 : $Tensor<Float>) [T $Float] : $Tensor<Float>
```
I'd suggest syntactically making the attributes another "parameter list" which is only printed if attributes are present.  The example above would be:

%2 = graph_op "tf.Add"(%0 : $Tensor<Float>, %1 : $Tensor<Float>) (T: $Float) : $Tensor<Float>

And if there are no attributes it would be omitted:

%2 = graph_op "tf.Add"(%0 : $Tensor<Float>, %1 : $Tensor<Float>) : $Tensor<Float>

WDYT?

rxwei replied:
Colons are going to be really confusing because the attribute value is not a type signature.

rxwei replied:
Plus, I think since `[]` is already used for instruction attributes, we can make it continue to carry that meaning.

dan-zheng replied:
I am a fan of the current syntax using `[]`.

- There is no colon between attribute names and values:
  - `[name1 1, name2 2.0, name3 $Float]`.
- If there are no attributes, the `[]` is not printed.

lattner replied:
As mentioned in another comment, we also have to add a (weird) type system for these constants as well.  I think that argues for using equals or something to distinguish these from normal arguments, e.g. something like:

[name1 = i64 1, name2 = f64 2.0, name3 = $Float, name4 = [[i32 1, i32 2, i32 3], f64 17.0]]

I still don't love the outer square brackets tho.

dan-zheng replied:
I handled integer/floating point datatypes, please re-review.
I'm not sure I parsed datatypes (`i32`, `f64`) in the optimal way, would appreciate your feedback.

I also added a `=` to separate attribute names from values: `<name> = <value>`.

lattner reviewed (commented):
This is really awesome, thank you for driving this!

dan-zheng committed:
Use `P.parseList` for `SymbolicValue` and `GraphOperationInst`, minor cleanup.

- Use `P.parseList` to parse `SymbolicValue` aggregates and `GraphOperationInst` arguments/attributes.
  - `parseList` disallows comma separator after last element in a list.
- Minor cleanup.

dan-zheng committed:
Parse datatype for integer/floating point `SymbolicValue`s.

Edit format for `GraphOperationInst` attributes. A `=` now separates name and
value: `<name> = <value>`.

Integer/floating point `SymbolicValue`s are now preceded by a datatype
(e.g. `name1 = i32 42` and `name2 = f64 3.0`).

Arbitrary integer bitwidths are supported.
For floating points, only the `f32` and `f64` bitwidths are supported.

lattner commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -7978,6 +7982,131 @@ inline DestructureTupleInst *DestructureTupleResult::getParent() {
+};
+
+/// SWIFT_ENABLE_TENSORFLOW
+/// An graph operation attribute. Attributes have a name and a constant value.
```
An graph -> A graph

dan-zheng replied:
Done, thanks.

lattner commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -885,6 +887,119 @@ void SILParser::convertRequirements(SILFunction *F,
+/// - An integer datatype and literal (i32 42).
+/// - A floating point datatype and literal (f64 3.14).
+/// - A metatype (the instance type is parsed) ($Float).
+/// - An aggregate ([1, 2, 3]).
```
Comment should be updated to include types, e.g.:  [i32 1, i64 2, f32 3]

dan-zheng replied:
I believe the comment already included types.
I added some more for extra clarity:
```diff
 // SWIFT_ENABLE_TENSORFLOW
 /// Parse a SIL constant value (one of the following categories):
-/// - An integer datatype and literal (i32 42).
-/// - A floating point datatype and literal (f64 3.14).
+/// - An integer datatype and literal (i32 -3, i64 42).
+/// - A floating point datatype and literal (f32 3, f64 3.14).
 /// - A metatype (the instance type is parsed) ($Float).
 /// - An aggregate ([1, 2, 3]).
 /// Returns true on error.
```

rxwei replied:
I saw you address my comment above by editing your comment. Marking to resolve.

lattner commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -885,6 +887,119 @@ void SILParser::convertRequirements(SILFunction *F,
+
+      // Parse floating point value.
+      double floatValue;
+      P.Tok.getText().getAsDouble(floatValue);
```
Is this what the float_literal instruction does?  You should follow its implementation for the parsing logic.

dan-zheng replied:
The `float_literal` instruction actually parses hex floating point values and prints the decimal representation in a comment:
`%5 = float_literal $Builtin.FPIEEE80, 0x4000C8F5C28F5C28F5C3 // 3.1400000000000000001`

Would you prefer that I use hexadecimal as well?
I couldn't think of a nice way to print decimal representations in a comment, so I just used decimal instead.

dan-zheng replied:
After discussion, we decided that:
* Parser should accept both decimal/hexadecimal formats.
* Printer should print hexadecimal only.

lattner commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -885,6 +887,119 @@ void SILParser::convertRequirements(SILFunction *F,
+    SourceLoc rSquareLoc;
+
+    SmallVector<SymbolicValue, 8> elements;
+    ParserStatus status =
```
nice!

lattner reviewed (approved):
Looks great, thank you!

dan-zheng committed:
Fix typos, update symbolic value textual representation comment.

dan-zheng committed:
Document `graph_op` instruction in SIL.rst.

dan-zheng committed:
Remove unused variable.

dan-zheng commented:
Two things before merging:
* How should floating point `SymbolicValue` be represented textually?
  * Currently, I'm using a decimal format (`3.1400000000000000001`), which is human-readable but inexact (parsed `3.14` is printed as `3.1400000000000000001`).
  * The alternative is to use a hexadecimal format (`0x4000C8F5C28F5C28F5C3`), following the `float_literal` instruction and LLVM IR. This is exact but not readable.
* I [documented `graph_op` in SIL.rst](https://github.com/apple/swift/pull/17077/commits/9a2a3129ca6d049e81582f3cab19a779f8066d69), please review.

rxwei commented:
I would recommend printing a comment on top of a `graph_op` inst by SIL printer, like how `function_ref` prints a demangled function in a comment above the instruction.

```
// "tf.Add" [T = $Float, shape = [f32 1.0, f32 2.0, i32 2]]
%add = graph_op "tf.Add"(%x : $Tensor<Float>, %y : $Tensor<Float>) [T = $Float] : $Tensor<Float>
```

dan-zheng commented:
Sure!
Did you intend to include `shape = [f32 1.0, f32 2.0, i32 2]` in the attribute list for `graph_op`?
`shape` appears in the comment but not the actual instruction.

rxwei commented:
Yep, that's what I meant by "... whatever hex".

lattner commented:
A comment on top of the instruction is going to be very complicated.  Please don't over design this.  It is fine to print:

f32 0x4000C8F5C28F5C28F5C3 /*1.0*/

rxwei commented:
Ok, inline comments look better.

dan-zheng committed:
Change floating point `SymbolicValue` textual representation.

- Parsing: both decimal and hexadecimal representations (in the style of
  `float_literal` instruction) are accepted.
- Printing: hexadecimal is printed, following by the decimal in an inline
  comment.

dan-zheng commented:
Please review: floating point textual representation changed in https://github.com/apple/swift/pull/17077/commits/9232753d9e255fada74f6e785f340297a0a2c989.

Parser accepts:
```
%0 = graph_op "tf.Dummy"() [hex_float = f64 0x40091EB851EB851F, dec_float = f32 3.14
```
Printer prints:
```
%0 = graph_op "tf.Dummy"() [hex_float = f64 0x40091EB851EB851F /* 3.1400000000000001 */, dec_float = f32 0x4048F5C3 /* 3.1400001 */
```

This is the last change planned before merging.

dan-zheng commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -937,21 +939,44 @@ static bool parseSymbolicValue(SymbolicValue &value, SILParser &SP,
 
       // Parse floating point value.
+      // Handle hexadecimal case.
+      if (P.Tok.is(tok::integer_literal)) {
```
Need to check that the integer literal is a hex string (e.g. starts with `0x`).

dan-zheng replied:
Done. Made other fixes in https://github.com/apple/swift/pull/17077/commits/6ab561d71284a8845eda1090959da586a247fa99 as well.

dan-zheng committed:
Parse negative floating point literals, add tests.

Make tests for parsing/printing `graph_op` instructions more robust.

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -963,20 +963,35 @@ static bool parseSymbolicValue(SymbolicValue &value, SILParser &SP,
       // Handle decimal case.
-      double floatValue;
-      P.Tok.getText().getAsDouble(floatValue);
+      double tmpValue;
```
Plz move this line right above its first use `P.Tok.getText().getAsDouble(tmpValue);`

dan-zheng replied:
Done.

dan-zheng committed:
Minor cleanup.

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
All tests pass.
@lattner, do things LGTY?

lattner commented:
Sorry for the last minute suggestion, but another project we're working on is grappling with a similar issue and came up with another syntax.  Instead of:

%1 = graph_op "tf.Dummy"() [hex1 = f64 0x40091EB851EB851F, hex2 = f32 0x4048F5C3] : $Tensor<Float>

How do you like this, connoting that the attributes are "dictionary like":

%1 = graph_op "tf.Dummy"() {hex1: f64 0x40091EB851EB851F, hex2: f32 0x4048F5C3} : $Tensor<Float>

This seems to have a nice consistency with the rest of swift.

rxwei commented:
Yeah I just saw that too. I like this syntax!

lattner reviewed (approved):
Aside from the {} syntax thing, the patch LGTM!

rxwei commented:
> Using colon is a bit problematic because of metatypes, as @rxwei previously stated.
[T: $Float] actually means T is $Float.Type.

Colon is only problematic in the context of parentheses, which makes it look like an argument list. I don't think there's any potential confusion within `{}`.

dan-zheng committed:
Update `graph_op` attribute list format.

The attribute list now uses a dictionary format: {<name>: <value>}.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Edit doc comments.

dan-zheng commented:
@mhong brought up a question: how should the device string attribute be handled?
It can either be a normal attribute in the attribute list (once String `SymbolicValue` is implemented), or it could be a special-case attribute on `graph_op` itself.

I'm leaning towards the first option. since there's no need to treat the "device" attribute specially.

mhong commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -885,6 +887,158 @@ void SILParser::convertRequirements(SILFunction *F,
+///   - The literal value may be in either decimal format or the hexadecimal
+///     format used by the 'float_literal' instruction.
+/// - A metatype (the instance type is parsed) ($Float).
+/// - An aggregate ([i32 1, i64 2, f32 3.0]).
```
what is the use case for "aggregate" here? also i didn't find its mention in the SIL.rst doc above.

dan-zheng replied:
Aggregates primarily represent constant arrays (e.g. tensor shapes).
They may also represent other aggregate types like structs or tuples.

Aggregate constants are useful for TensorFlow ops that have an array attribute, such as [`Conv2D`](https://github.com/tensorflow/tensorflow/blob/694b75a24bce416555425dedc58b0cdcd0d52c1e/tensorflow/core/ops/nn_ops.cc#L265):
```c++
REGISTER_OP("Conv2D")
    .Input("input: T")
    .Input("filter: T")
    .Output("output: T")
    .Attr("T: {half, bfloat16, float, double}")
    .Attr("strides: list(int)")
    .Attr("use_cudnn_on_gpu: bool = true")
    .Attr(GetPaddingAttrString())
    .Attr(GetConvnetDataFormatAttrString())
    .Attr("dilations: list(int) = [1, 1, 1, 1]")
    .SetShapeFn(shape_inference::Conv2DShape);
```

I'll add more information about aggregates in the doc comment.

dan-zheng committed:
Verify that `graph_op` attribute names are unique.

mhong commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -885,6 +887,158 @@ void SILParser::convertRequirements(SILFunction *F,
+  return true;
+};
+
 /// SWIFT_ENABLE_TENSORFLOW
```
this tag can be consolidated with the above code?

dan-zheng replied:
I would prefer to keep multiple tags so that each function is marked clearly.

mhong commented on test/TensorFlow/graph_op_inst.sil:
```diff
@@ -0,0 +1,69 @@
+// RUN: %target-sil-opt -sil-print-all %s | %FileCheck %s
```
also add some negative test cases?
e.g. P.diagnose(P.Tok, diag::expected_tok_in_sil_instr, "graph_op name");

dan-zheng replied:
There isn't a tradition of testing SIL parser errors, unfortunately. I'm not able to write an effective unit test because introducing one error causes many unrelated parser errors to occur.
SIL verification handles errors instead.

mhong replied:
Ok. Agree that's unfortunate.

mhong reviewed (commented):
What a great addition to our infrastructure!

mhong commented on lib/SIL/SILVerifier.cpp:
```diff
@@ -1188,8 +1188,12 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
-    for (auto attr : GI->getAttributes())
+    llvm::DenseSet<Identifier> attributeNames;
+    for (auto attr : GI->getAttributes()) {
+      require(attributeNames.insert(attr.name).second,
```
add a test case?

rxwei replied:
There's no precedent of test-casing SIL verification cuz each `require` failure is a trap. There's only examples like testing whether `sil-opt` crashes or not (see [example](https://github.com/apple/swift/blob/master/test/SIL/verifier-fail-instruction.sil)). Doing so for every single `require` instruction is not practical. OTOH, IR verification is an implementation  detail anyway.

dan-zheng commented on include/swift/SIL/SILBuilder.h:
```diff
@@ -1278,6 +1278,15 @@ class SILBuilder {
 
+  GraphOperationInst *createGraphOperation(SILLocation loc, Identifier name,
+                                           ArrayRef<SILValue> operands,
+                                        ArrayRef<GraphOperationAttribute> attrs,
```
The indentation is intentional to maintain 80-column limit.

mhong replied:
A more common wrapping in this case would be:

GraphOperationInst *createGraphOperation(
  param1,
  param2,
  param3);

I'm also ok with the following (clang-format does that for me automatically):
GraphOperationInst *createGraphOperation(
  param1, param2,
  param3);  // this assumes param3 is too long to put in the previous line.

lattner commented:
Agreed to comments above: we should definitely support string literal, but doing so in a future patch is totally reasonable.

dan-zheng commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -7978,6 +7982,131 @@ inline DestructureTupleInst *DestructureTupleResult::getParent() {
+  }
+};
+
+// SWIFT_ENABLE_TENSORFLOW
```
I would prefer to keep it, as it marks a newly added function.

mhong replied:
SG

dan-zheng commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -7978,6 +7982,131 @@ inline DestructureTupleInst *DestructureTupleResult::getParent() {
+
+  GraphOperationInst *getParent();
+  const GraphOperationInst *getParent() const {
+    return const_cast<GraphOperationResult *>(this)->getParent();
```
Do you have a concrete suggestion in mind?
All other subclasses of `MultipleValueInstructionResult` implement `getParent` in this way.

dan-zheng committed:
Minor edits.

Address comments from @mhong.

dan-zheng commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -7978,6 +7982,131 @@ inline DestructureTupleInst *DestructureTupleResult::getParent() {
+    return N->getKind() == SILNodeKind::GraphOperationResult;
+  }
+
+  GraphOperationInst *getParent();
```
Thanks for the catch, I inlined the function.
I didn't inline it originally because I copied the code from [`DestructureTupleInst`](https://github.com/dan-zheng/swift/blob/534810d232e6c2337bc824b6745f53f14c2385bb/include/swift/SIL/SILInstruction.h#L7980).

dan-zheng commented:
@lattner or @rxwei: could you please do a final review before I merge the PR?
Since your last review, the main change is the attribute list syntax (it's now dictionary-like, `{T: $Float}`).

mhong commented:
Previously user can write #tfop(...) directly. With this new inst type, is it still possible? Having such support as an "escape hatch" could be useful, but it depends on the eng feasibility and cost.

If user can write this directly in some way, and user can introduce a bad inst (say with duplicate attrs), then we'll need to add diagnostics code to reject such code, instead of just crashing during SIL verification. For eng simplicity, it'd be nice if user can never generate such bad insts, so that the problem of rejecting them can be defined away.

dan-zheng commented:
`#tfop` is the AST-level representation of TensorFlow operations and it's not affected by this PR.
The next step after this PR is to change the lowered representation of `#tfop` from `BuiltinInst` to `GraphOperationInst`.
 
Currently, `#tfop` is represented as an `ObjectLiteralExpr` but it should eventually be rewritten as a new expression (something like `GraphOperationExpr`). New sema rules for `#tfop` should be added to check for duplicate attributes. This guarantees that lowered `GraphOperationInst`s will be valid.

mhong commented:
The plan sounds good!

rxwei commented:
> Previously user can write #tfop(...) directly. With this new inst type, is it still possible? Having such support as an "escape hatch" could be useful, but it depends on the eng feasibility and cost.

There is no syntax changes on the Swift side. This change is SIL-only. 

> Currently, #tfop is represented as an ObjectLiteralExpr but it should eventually be rewritten as a new expression (something like GraphOperationExpr). New sema rules for #tfop should be added to check for duplicate attributes. This guarantees that lowered GraphOperationInsts will be valid.

The initial plan is to make deabstraction turn all builtin tfops to this new instruction, so that partitioning and graph lowering can deal with attributes in-line instead of decoding strings and gather data flow dependencies. This will also greatly simplify SIL cloning in the partitioner. `graph_op` is not directly emitted by SILGen. Instead, it is an intermediate instruction between deabstraction and partitioning/graphgen.

New changes LGTM. Feel free to merge!

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
Merged. Thanks everyone for your careful review and feedback!
